site/content/posts/2023-09-18-multiple-idps-and-identity-transformations.md

```diff
@@ -0,0 +1,127 @@
+that the user should be allowed to continue their ongoing session.
+
+Prior to this release, there was no way to configure each FederationDomain to be meaningfully different from the others
+within a single Pinniped Supervisor. However, starting with this release, you can now configure each FederationDomain
+to allow a different pool of users. (See below for more details.) This makes it possible to use multiple FederationDomains
+in a single Pinniped Supervisor.
+
+## New: Multiple identity providers
+
+This release adds a new `spec.identityProviders` configuration option to the FederationDomains resource. Each FederationDomain
+may choose which OIDCIdentityProviders, LDAPIdentityProviders, and ActiveDirectoryIdentityProviders users may use to
+authenticate into the clusters which trust that FederationDomain.
+
+Why would an administrator want to use multiple identity providers? Here are some examples:
+- You might like to configure multiple FederationDomains, each using a different identity provider.
+  Each FederationDomain can be used to provide authentication to a group of Kubernetes clusters.
+  For example, you could configure a FederationDomain for each division of your R&D department,
+  thus preventing developers of each division from using each other's clusters.
+- Within a single FederationDomain, you might like to use one identity provider for admin-level
+  users, while using another identity provider for developer-level users, thus allowing you to draw identities for
+  different roles from different user databases.
+- Within a single FederationDomain, you might like to allow users from two different organizations to both
+  be able to authenticate into the same clusters by configuring an identity provider for each. This could allow
+  multiple teams to collaborate on the same clusters.
+- Within a single FederationDomain, you might like to use the same external identity provider configured
+  different ways to get different types of users. For example, you could configure one OIDCIdentityProvider resource
+  to use a client ID for your human users which requires them to authenticate using multi-factor
+  authentication (configured as a setting on that client ID), while configuring another OIDCIdentityProvider
+  (with a different client ID) to allow CI/CD "bot" users to authenticate with a single factor to ease automation.
+- You might like to configure multiple LDAPIdentityProviders to each search for users or groups in a different branch
+  of your LDAP tree. Then you could use each LDAPIdentityProvider in a different FederationDomain, or you could
+  use multiple within the same FederationDomain.
+
+Configuring a FederationDomain to use one or more identity providers is as easy as listing `objectRefs`
+in the FederationDomain's spec for each OIDCIdentityProviders, LDAPIdentityProviders, or ActiveDirectoryIdentityProviders
+that you wish to use in the FederationDomain.
+
+## New: Identity transformations and policies
```

Comment:
I wonder if we should reinforce here that Pinniped *WILL NOT* automatically provide transformations or guard against any side effects such as duplicate usernames, unintended groups leaking into tokens, etc.

Reply:
(unintended groups leaking into tokens: someone could figure out that "admin" in IDP A grants kubernetes admin privileges and so they grant themselves "admin" in IDP B even though that really means something different).

Reply:
I added this new parenthetical note to the relevant paragraph:

> (Note that Pinniped will not automatically avoid these username and group name conflicts. You must explicitly configure identity transformations to add prefixes or otherwise resolve conflicting names.)

In terms of unintended groups leaking into tokens, Pinniped has always included all your group names into ID tokens (or at least all the group names that it was configured to find), so that's not new. It is new that there is now a way that you can further filter down those group names using a transformation, but I mentioned that as an example of something that you can do with transformations where I said:

> Drop groups from a user's list of groups which are not interesting for your Kubernetes authorization use cases

Reply:
We had discussed this previously.  One idea I am still fond of is making a transformation pipeline `REQUIRED`.  However, its fine if a user configures a `noop`.  The idea is that a user has to opt-in and say "yes, I know I am getting exactly the identities with no transformation provided from these IDPs", which would at least require a first look at the tranformations provided.  It is slight friction, but not a ton.  I'd rather that slight friction to remind users that this feature exists vs an accidental security issue due to "oops".

Reply:
@benjaminapetersen We all decided not to do that. The user experience is weird, and it doesn't add any security.

Reply:
I do vaguely recall the conversation, but I did not remember a final decision being made.